library/src/image.cxx

```diff
@@ -351,6 +353,149 @@ std::vector<unsigned char> image::saveBuffer(SaveFormat format) const
   }
 }
 
+//----------------------------------------------------------------------------
+const f3d::image& image::toTerminalText(std::ostream& stream) const
+{
+  const int depth = this->getChannelCount();
+  if (this->getChannelType() != ChannelType::BYTE || depth < 3 || depth > 4)
+  {
+    throw std::invalid_argument("image must be byte RGB or RGBA");
+  }
+
+  int dims[3];
+  this->Internals->Image->GetDimensions(dims);
+  const int width = dims[0];
+  const int height = dims[1];
+  const unsigned char* content = static_cast<unsigned char*>(this->getContent());
+
+  constexpr unsigned char alphaCutoff = 127;
+
+  /* Function to retrieve pixels so we can return:
+    - transparent black values for out-of-bounds coords,
+    - opaque alpha value when the image has no alpha channel.
+    Rendering with half blocks means 1 line of text represents 2 rows of pixels
+    so we _will_ attempt to access a line past the bottom if the height is not even.
+  */
+  const auto getPixel = [=](int x, int y)
+  {
+    if (x >= 0 && x < width && y >= 0 && y < height)
+    {
+      const size_t i = depth * ((height - 1 - y) * width + x);
+      const int rgb = content[i + 0] << 16 | content[i + 1] << 8 | content[i + 2];
+      const bool transparent = depth > 3 ? content[i + 3] <= alphaCutoff : false;
+      return std::make_pair(rgb, transparent);
+    }
+    return std::make_pair(0x000000, true);
+  };
+
+  /* Functions to manipulate the terminal colors using escape sequences.
+    Keep track of the foreground and background states to avoid redundant sequences.
+  */
+  int currentFg = -1;
+  int currentBg = -1;
+  const auto setFg = [&stream, &currentFg](int rgb)
```

Comment:
Since everything is captured by reference, let's simplify that to
```suggestion
  const auto setFg = [&](int rgb)
```
(and the other ones below)